pkg/csi/manila/controllerserver.go

```diff
@@ -248,6 +228,33 @@ func (cs *controllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
 			sourceShare.ShareProto, req.GetSourceVolumeId(), cs.d.shareProto)
 	}
 
+	// In order to satisfy CSI spec requirements around CREATE_DELETE_SNAPSHOT
+	// and the ability to populate volumes with snapshot contents, share type
+	// of the source share must advertise create_share_from_snapshot_support.
+	//
+	// TODO: It's not enough to validate only share type caps as they can change
+	//       over time. create_share_from_snapshot_support capability of the source
+	//       share needs to be checked as well. Add this check once this is exposed
+	//       in gopherclouds shares.Share struct.
+
+	shareTypeCaps, err := capabilities.GetManilaCapabilities(sourceShare.ShareType, manilaClient)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to get Manila capabilities for share type %s: %v", sourceShare.ShareType, err)
+	}
+
+	requiredShareTypeCaps := []capabilities.ManilaCapability{
+		capabilities.ManilaCapabilitySnapshot,
+		capabilities.ManilaCapabilityShareFromSnapshot,
+	}
+
+	for _, c := range requiredShareTypeCaps {
+		if !shareTypeCaps[c] {
+			return nil, status.Errorf(codes.InvalidArgument,
```

Comment:
The main objective here is to ensure that it will be possible to create a new volume with this snapshot as its source and I'm not sure I'm going the right way about this.

Is it possible to create a share from a snapshot, but using a share type that's different than what was used to create the original share? If so, then checking share type caps here doesn't make much sense -- because maybe one type had `create_share_from_snapshot_support` disabled, but the other may have it enabled and share creation would succeed.

From my tests, it seems the caps on the share (not on the share type) seemed to have the final say. Even if I tried to create a share from snapshot with a type that had `create_share_from_snapshot_support=True`, Manila would refuse to create it if the original share reported `create_share_from_snapshot_support=False`. Unfortunately, there's no way for manila-csi to know at the moment. The gophercloud lib is currently missing `create_share_from_snapshot_support` field in [`shares.Share` struct](https://github.com/gophercloud/gophercloud/blob/932f0916e113975e4d34bcd1d2dce9c01aa96570/openstack/sharedfilesystems/v2/shares/results.go#L18-L76). I'll push a patch there. I'm not entirely sure we should merge this PR without it.

Reply:
https://github.com/gophercloud/gophercloud/pull/2191

Reply:
@gman0: no you can't specify both snapshot and share type when creating new shares; new shares from snapshots are expected to retain the share type of the parent; 

and yes, the capability of the share is the best place to look since share type extra specs are mutable. While share type extra-specs are mutable, they're assumed to not change often and manila administrators know that changes aren't going to apply to resources that are already provisioned. However, i totally agree with not relying on the share type extra-specs. 

your proposed way of just checking the share's "snapshot_support" and "create_share_from_snapshot_support" seems correct.

Reply:
@gouthampacha thank you for explanation and suggestions! I'll fix this like we've agreed after the `create_share_from_snapshot_support` patch in gophercloud is in.